src/coreclr/src/vm/weakreferencenative.cpp

```diff
@@ -132,21 +133,40 @@ IWeakReference* GetWinRTWeakReference(OBJECTREF* pObject)
 
     MethodTable* pMT = (*pObject)->GetMethodTable();
 
-    // If the object is not an RCW, then we do not want to use a WinRT weak reference to it
-    if (!pMT->IsComObjectType())
-    {
-        return nullptr;
-    }
+    IWeakReferenceSource* pWeakReferenceSourceUnsafe;
 
-    // If the object is a managed type deriving from a COM type, then we also do not want to use a WinRT
+    // If the object is not an RCW, then we do not want to use a native COM weak reference to it
+    // If the object is a managed type deriving from a COM type, then we also do not want to use a native COM
     // weak reference to it.  (Otherwise, we'll wind up resolving IWeakReference-s back into the CLR
     // when we don't want to have reentrancy).
-    if (pMT != g_pBaseCOMObject && pMT->IsExtensibleRCW())
+    if (pMT->IsComObjectType()
+     && (pMT == g_pBaseCOMObject || !pMT->IsExtensibleRCW()))
     {
-        return nullptr;
+        pWeakReferenceSourceUnsafe = reinterpret_cast<IWeakReferenceSource*>(GetComIPFromObjectRef(pObject, IID_IWeakReferenceSource, false /* throwIfNoComIP */));
     }
+#ifdef FEATURE_COMWRAPPERS
+    else
+    {
+        SyncBlock* syncBlock = (*pObject)->GetSyncBlock();
+        InteropSyncBlockInfo* interopInfo = syncBlock->GetInteropInfo();
+        void* context;
+        if (!interopInfo->TryGetExternalComObjectContext(&context))
+        {
+            return nullptr;
+        }
```

Comment:
Push all of this manipulation into the interoplibinterface contract. Knowing how we discover the associated code isn't interesting here. The call should be something like `static IUnknown* GetIdentityForObject(_In_ OBJECTREF* context);`

This will return `nullptr` or `IUnknown*`.